suricata/update/parsers.py

```diff
@@ -0,0 +1,164 @@
+    update_parser.add_argument("--etopen", action="store_true",
+                               help="Use ET-Open rules (default)")
+    update_parser.add_argument("--reload-command", metavar="<command>",
+                               help="Command to run after update if modified")
+    update_parser.add_argument("--no-reload", action="store_true", default=False,
+                               help="Disable reload")
+    update_parser.add_argument("-T", "--test-command", metavar="<command>",
+                               help="Command to test Suricata configuration")
+    update_parser.add_argument("--no-test", action="store_true", default=False,
+                               help="Disable testing rules with Suricata")
+
+    update_parser.add_argument("--no-merge", action="store_true", default=False,
+                               help="Do not merge the rules into a single file")
+
+    # Hidden argument, --now to bypass the timebased bypass of
+    # updating a ruleset.
+    update_parser.add_argument(
+        "--now", default=False, action="store_true", help=argparse.SUPPRESS)
+
+    update_parser.epilog = r"""other commands:
+    update-sources             Update the source index
+    list-sources               List available sources
+    enable-source              Enable a source from the index
+    disable-source             Disable an enabled source
+    remove-source              Remove an enabled or disabled source
+    list-enabled-sources       List all enabled sources
+    add-source                 Add a new source by URL
+"""
+
+    # The Python 2.7 argparse module does prefix matching which can be
+    # undesirable. Reserve some names here that would match existing
+    # options to prevent prefix matching.
+    update_parser.add_argument("--disable", default=False,
+                               help=argparse.SUPPRESS)
+    update_parser.add_argument("--enable", default=False,
+                               help=argparse.SUPPRESS)
+    update_parser.add_argument("--modify", default=False,
+                               help=argparse.SUPPRESS)
+    update_parser.add_argument("--drop", default=False, help=argparse.SUPPRESS)
+
+    commands.listsources.register(subparsers.add_parser(
+        "list-sources", parents=[global_parser]))
+    commands.listenabledsources.register(subparsers.add_parser(
+        "list-enabled-sources", parents=[global_parser]))
+    commands.addsource.register(subparsers.add_parser(
+        "add-source", parents=[global_parser]))
+    commands.updatesources.register(subparsers.add_parser(
+        "update-sources", parents=[global_parser]))
+    commands.enablesource.register(subparsers.add_parser(
+        "enable-source", parents=[global_parser]))
+    commands.disablesource.register(subparsers.add_parser(
+        "disable-source", parents=[global_parser]))
+    commands.removesource.register(subparsers.add_parser(
+        "remove-source", parents=[global_parser]))
+
```

Comment:
Could you try and break this into smaller functions? Also, there is a lot of repetition of `add_argument` where a loop could have been used. Could you come up with a solution to make the code more compact? Let's discuss here itself. Thanks.

Reply:
Apologies for the late reply. 
To make the code more compact and to reduce the repetition of `add_argument`, I tried to store the arguments along with their named options (like help, action etc.) in a tuple and add them to the parser in a loop. Since, syntax of the add_argument() function expects a list of flags as the first arguments, I used tuple unpacking ([this](https://docs.python.org/2/tutorial/controlflow.html#unpacking-argument-lists)) for the positional arguments. I tried something like this:
```
global_parser = argparse.ArgumentParser(add_help=False)

    arguments = [
    (("-v", "--verbose"),
        {'action': 'store_true', 'default': 'None',
            'help': "Be more verbose" } ),
    (("-q", "--quiet"),
        {'action': 'store_true', 'default': 'None',
            'help': "Be quiet, warning and error messages only" } ),
    (("-D", "--data-dir"),
        {'metavar': '<directory>', 'dest': 'data_dir',
            'help': "Data directory (default: /var/lib/suricata)" } ),
    (("-c", "--config"),
        {'metavar': '<filename>',
            'help': "configuration file (default: %s)" %(default_update_yaml) } ),
    (("--suricata-conf",),
        {'metavar': '<filename>',
            'help': "configuration file (default: /etc/suricata/suricata.yaml)" } ),
    (("--suricata",),
        {'metavar': '<path>',
            'help': "Path to Suricata program" } ),
    (("--suricata-version",),
        {'metavar': '<version>',
            'help': "Override Suricata version" } ),
    (("--user-agent",),
        {'metavar': '<user-agent>',
            'help': "Set custom user-agent string" } ),
    (("--no-check-certificate",),
        {'action': 'store_true', 'default': 'None',
            'help': "Disable server SSL/TLS certificate verification" } ),
    (("-V", "--version"),
        {'action': 'store_true', 'default': 'False',
            'help': "Display version" } )
    ]

    for arg, opts in arguments:
        global_parser.add_argument(*arg, **opts)
```
It seems to work fine. Does it look good to you? 
Although I haven't yet figured how to break the parsers.py code further into smaller functions. Any input would be helpful :)

Reply:
Seems nice. One way of breaking up the parsers functions could be based on the different parsers in the function. e.g. there is update parser, global parser and then there are commands.